### PR TITLE
Jerome dev

### PR DESCRIPTION
**Pull request by jeromel05**

### CODE REVIEW

File: dinov2/configs/train/whoi.yaml [hunk 1]
```diff
@@ -60,21 +60,24 @@ ibot:
...
-  saveckp_freq: 15
+  saveckp_freq: 10
   seed: 0
-  num_workers: 4
-  OFFICIAL_EPOCH_LENGTH: 1250
-  cache_dataset: true
+  num_workers: 8
+  OFFICIAL_EPOCH_LENGTH: 2500
   centering: "sinkhorn_knopp" # or "centering"
+  cache_dataset: false
```
Comment (NoraKrbr) [inline dinov2/configs/train/whoi.yaml:70]: why is the dataset not cached?

File: dinov2/data/collate.py [hunk 2]
```diff
@@ -7,15 +7,27 @@
...
     # dtype = torch.half  # TODO: Remove
 
-    n_global_crops = len(samples_list[0][0]["global_crops"])
-    n_local_crops = len(samples_list[0][0]["local_crops"])
-
-    collated_global_crops = torch.stack([s[0]["global_crops"][i] for i in range(n_global_crops) for s in samples_list])
-
-    collated_local_crops = torch.stack([s[0]["local_crops"][i] for i in range(n_local_crops) for s in samples_list])
+    # samples dict_keys(['global_crops', 'global_crops_teacher', 'local_crops', 'offsets'])
+    # TODO: use global_crops_teacher in returned dict
```
Comment (NoraKrbr) [inline dinov2/data/collate.py:14]: not TODO anymore, right?

File: dinov2/data/datasets/extended.py [hunk 3]
```diff
@@ -3,10 +3,14 @@
 # This source code is licensed under the Apache License, Version 2.0
 # found in the LICENSE file in the root directory of this source tree.
 
+from typing import Union
```
Comment (NoraKrbr) [inline dinov2/data/datasets/extended.py:6]: merge lines 6 and 8: `from typing import import Any, Tuple. Union`

File: dinov2/data/datasets/extended.py [hunk 4]
```diff
@@ -22,22 +26,30 @@ def get_image_data(self, index: int) -> bytes:
...
+                image = ImageDataDecoder(img_bytes).decode()
+            except Exception as e:
+                raise RuntimeError(f"can not read image for sample {index}") from e
+
         target = self.get_target(index)
         target = TargetDecoder(target).decode()
 
         if self.transforms is not None:
             image, target = self.transforms(image, target)
+            return image, target
```
Comment (NoraKrbr) [inline dinov2/data/datasets/extended.py:49]: `return` statement here is the same in `if` and `else`, can be moved outside of the conditional?

File: dinov2/eval/segmentation/models/decode_heads/linear_head.py [hunk 5]
```diff
@@ -8,8 +8,8 @@
 
 from mmseg.models.builder import HEADS
 from mmseg.models.decode_heads.decode_head import BaseDecodeHead
-from mmseg.ops import resize
-
+from mmcv.transforms import Resize
```
Comment (NoraKrbr) [inline dinov2/eval/segmentation/models/decode_heads/linear_head.py:11]: `Resize` is not used, remove from imports?

File: dinov2/train/train.py [hunk 6]
```diff
@@ -205,45 +251,91 @@ def do_train(cfg, model, resume=False):
...
+        activities = [ProfilerActivity.CPU, ProfilerActivity.CUDA]
+        profiler_dir = os.path.join(cfg.train.output_dir, "profiler")
+        os.makedirs(profiler_dir, exist_ok=True)
+        profiler = torch.profiler.profile(
+            schedule=profiler_schedule,
+            activities=activities,
+            on_trace_ready=torch.profiler.tensorboard_trace_handler(profiler_dir),
+            with_stack=True,
+        )
+        profiler.start()
```
Comment (NoraKrbr) [inline dinov2/train/train.py:309]: does profiling work now? Otherwise, maybe we should not include it in main yet?

File: dinov2/configs/train/whoi.yaml [hunk 7]
```diff
@@ -60,21 +60,24 @@ ibot:
...
-  dataset_path: HDF5Dataset:split=TRAIN:root=/fast/AG_Kainmueller/plankton/data/WHOI/preprocessed_hdf5:extra=*
+  dataset_path: LMDBDataset:split=ALL:root=/fast/AG_Kainmueller/plankton/data/WHOI/preprocessed/preprocessed_lmdb/:extra=*
   output_dir: 'home/jluesch/output_dir'
-  saveckp_freq: 15
+  saveckp_freq: 10
   seed: 0
-  num_workers: 4
-  OFFICIAL_EPOCH_LENGTH: 1250
-  cache_dataset: true
+  num_workers: 8
```
[the remark this answers is not in the document]
Comment (jeromel05) [reply dinov2/configs/train/whoi.yaml:67]: Num_workers 8 or 16 is the fastest

File: dinov2/configs/train/whoi.yaml (hunk 1 above)
Comment (jeromel05) [reply dinov2/configs/train/whoi.yaml:70]: Caching the dataseyt did not show any speed improvements until now, so leaving default (not cached)

File: dinov2/data/collate.py (hunk 2 above)
Comment (jeromel05) [reply dinov2/data/collate.py:14]: Yes :)

File: dinov2/data/datasets/extended.py (hunk 3 above)
Comment (jeromel05) [reply dinov2/data/datasets/extended.py:6]: Done

Comment (jeromel05) [reply dinov2/data/datasets/extended.py:6]: Use Ruff imports to harmonize all files

File: dinov2/data/datasets/extended.py (hunk 4 above)
Comment (jeromel05) [reply dinov2/data/datasets/extended.py:49]: Good point, done

File: dinov2/train/train.py (hunk 6 above)
Comment (jeromel05) [reply dinov2/train/train.py:309]: I think it should be included because I am confident that the code is correct (copied from tutorial), and this code should work once we upgrade to PyTorch 2.1

**Comment by jeromel05** [comment]
Profiling, lmdb dataset, Ckpt saving on CPU, augmentation select (GPU or CPU), Ruff format, Bypass PIL Image and go to Tensor directly, use teacher crops, Interpolate positional embeddings, Safer load pretrained weights, Use whoi dataset means, torchvision transforms v2, torch compile, more metrics, print number of shards, add ALL split, fix Koleo loss to inf
